core/kernel/tee_ta_manager.c

```diff
@@ -261,6 +262,89 @@ static void tee_ta_unlink_session(struct tee_ta_session *s,
 	mutex_unlock(&tee_ta_mutex);
 }
 
+static void destroy_session(struct tee_ta_session *s,
+			    struct tee_ta_session_head *open_sessions)
+{
+	tee_ta_unlink_session(s, open_sessions);
+#if defined(CFG_TA_GPROF_SUPPORT)
+	free(s->sbuf);
+#endif
+	free(s);
+}
+
+static void destroy_context(struct tee_ta_ctx *ctx)
+{
+	DMSG("Destroy TA ctx (0x%" PRIxVA ")",  (vaddr_t)ctx);
+
+	condvar_destroy(&ctx->busy_cv);
+	pgt_flush_ctx(ctx);
+	ctx->ops->destroy(ctx);
+}
+
+static void destroy_ta_ctx_from_session(struct tee_ta_session *sess)
+{
+	struct tee_ta_session *ca_sess = NULL;
+	struct tee_ta_session *ta_sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct user_ta_ctx *utc = NULL;
+	size_t count = 1; /* start counting the references to the context */
+
+	mutex_lock(&tee_ta_mutex);
+
+	entry_get_open_sessions(&open_sessions);
+
+	DMSG("Remove any reference to context (0x%" PRIxVA ")",
+					      (vaddr_t)sess->ctx);
+
+	/*
+	 * Look over the entire list of opened sessions from normal world
+	 * for those having the same context and remove any reference to it.
+	 */
+	TAILQ_FOREACH(ca_sess, open_sessions, link) {
+		if (!ca_sess->ctx) /* the reference has been already removed */
+			continue;
```

Comment:
On the contrary, this line makes sure that no null ctx setting in the session data structure (ignore the sessions whose context reference has already been removed) is interpreted any further. This case was revealed by the tests described in the [comment](https://github.com/OP-TEE/optee_test/pull/335#issuecomment-474460022) and present in the updated version of [PR#355](https://github.com/OP-TEE/optee_test/pull/335).

Without it the execution hangs at the line:

`utc = to_user_ta_ctx(ca_sess->ctx);`

Do you think that we should separate the scanning loop for TA-to-TA and call it from tee svc instead?

Reply:
You're right, this test here `if (!ca_sess->ctx) continue;` is fine. Do not modify it.

But, this loop does not cover all TAs and sessions.
It scans all sessions opened from all non-secure clients and all sessions opened by a TA invoked by a non-secure client.

Consider the following scenario. Non-secure client CA opens a session over TA1 which invokes TA2 which invokes TA3. Then CA opens a session over TA3, which panics. The current loop will not see TA2->TA3 session.

If think this function need to scan all non-secure-to-TA session (as done here) and all TA contexts (to scan their opened session). `TAILQ_FOREACH(ctx, &tee_ctxes, link)` from `tee_ta_manager.c` scans all existing TA contexts.

Reply:
Just to make sure we are on the same page. You mentioned the case of chaining CA->TA1->TA2->TA3 and pointed out that this current implementation is missing TA2->TA3 sesssion. Moreover this could be overcome by scanning every context in `tee_ctxes` for its opened sessions.

I might have couple of questions relating this latest sentence:

1.  How can we relate the opened sessions to a specific context? I did the other way around: scan for open sessions and check in their internal structure if ctx matched. I have to admit that this last topic ran me into troubles. I would appreciate few more help here.

2. I thought is just enough to remove the panicked context from `tee_ctxes` as it is done like

`TAILQ_REMOVE(&tee_ctxes, sess->ctx, link);`

Reply:
> How can we relate the opened sessions to a specific context? I did the other way around: scan for open sessions and check in their internal structure if ctx matched. I have to admit that this last topic ran me into troubles. I would appreciate few more help here.

I think you did it the right way. Your sequence for releasing the context memory expect this loop to scan all session instance to nullify the TA context reference, but for the session that made the TA panicking.

To scan all sessions, I strongly think you need to:
* scan all sessions opened from non-secure, that is the current change:
```
	nsec_sessions_list_head(&open_sessions);
	TAILQ_FOREACH(ca_sess, open_sessions, link) {
		if (ca_sess->ctx == sess->ctx && ca_sess != sess) {
			ca_sess->ctx = NULL;
			count++;
		}
	}
```
* scan all sessions opened from a TA, that is scanning all TA contexts and, for each, scan all opened session, that is something like:
```
	TAILQ_FOREACH(ta_ctx, &tee_ctxes, link) {
		TAILQ_FOREACH(ta_sess, to_user_ta_ctx(ta_ctx)->open_sessions, link) {
			if (ta_sess->ctx == sess->ctx && ta_sess != sess) {
				ta_sess->ctx = NULL;
				count++;
			}
		}
	}
```

Reply:
When a PTA panics, it call `panic()`, no TA context to destroy, rather shutdown or reboot. PTAs are not applications. PTAs execute in OP-TEE core privileged context.
So you don't have to bother with. Nicely ignore their context and its fine, in my opinion.